### DISCUSSION TITLE
Path to resource is getting its backslashes removed on WIndows

### DISCUSSION THREAD

**Discussion by LFDanLu** [webpack/webpack#18681] [Q&A]
Hi, I've been running into a issue where it seems like a path to a resource is losing its backslashes on Windows. It seems like a classic escaping issue but attempts to work around/fix the path haven't worked out. This very well maybe a fault with something in https://github.com/devongovett/unplugin-parcel-macros but I've hit a wall debugging walking through the stack trace and would appreciate some pointers.

To reproduce: 
repository: https://github.com/LFDanLu/s2-test-app
Using Git Bash on Windows. node version 18.18.2
1. Pull the project down and install.
2. Run using yarn start.
3. Note the follow error appears:
```
ERROR in C:Usersdl164OneDriveDocumentsdevempty-projectsrcApp.js.macro-fb0d4180853462d3bcb94480fb3b2b62c62519e0b35074e4a51d1c603706c075.css
Module build failed: UnhandledSchemeError: Reading from "C:Usersdl164OneDriveDocumentsdevempty-projectsrcApp.js.macro-fb0d4180853462d3bcb94480fb3b2b62c62519e0b35074e4a51d1c603706c075.css" is not handled by plugins (Unhandled scheme).
Webpack supports "data:" and "file:" URIs by default.
You may need an additional plugin to handle "c:" URIs.
```

Locally, I've confirmed that the output of the [unplugin-macro](https://github.com/devongovett/unplugin-parcel-macros/blob/d523d9f7c768050e83e5bedc503df75377a3d68a/unplugin-macros.js#L132) returns code that looks something like
```
  code: 'import "@react-spectrum/s2/page.css";\n' +
    'import { Button } from "@react-spectrum/s2";\n' +
    'import React from "react";\n' +
    "import { hot } from 'react-hot-loader/root';\n" +
    'class App extends React.Component {\n' +
    '    render() {\n' +
    '        const { name } = this.props;\n' +
    '        return (<>\n' +
    '\n' +
    '      <div>awegwge</div>\n' +
    '\n' +
    '      <Button styles={" . ye"}>\n' +
    '\n' +
    '        Hello Spectrum 2!\n' +
    '\n' +
    '      </Button>\n' +
    '\n' +
    '      </>);\n' +
    '    }\n' +
    '}\n' +
    'export default hot(App);\n' +
    '\n' +
    'import "C:\\Users\\dl164\\OneDrive\\Documents\\dev\\empty-project\\src\\App.js.macro-fb0d4180853462d3bcb94480fb3b2b62c62519e0b35074e4a51d1c603706c075.css";',
  map: '{"version":3,"sources":["test.js"],"names":[],"mappings":"AAAA,OAAO,8BAA8B;AAErC,SAAS,MAAM,QAAQ,qBAAqB;AAC5C,OAAO,WAAW,QAAQ;AAC1B,SAAS,GAAG,QAAQ,wBAAwB;AAE5C,MAAM,YAAY,MAAM,SAAS;IAC/B,SAAS;QACP,MAAM,EAAE,IAAI,EAAE,GAAG,IAAI,CAAC,KAAK;QAC3B,OAAO,GACH;;MACF,CAAC,IAAI,OAAO,EAAE,IAAI;;MAClB,CAAC,OACC,iBAGD;;;;MAED,EAAE,OAAO;;MACT,GACF;IACF;AACF;AAEA,eAAe,IAAI,KAAK"}'
```
where it looks like the path is properly escaped. However, as I stepped through the stack I eventually ended up in Compilation where the request field in the [dependencies ](https://github.com/webpack/webpack/blob/46cee287199d7dce6879aa78046fae1df9a101e6/lib/Compilation.js#L2051)shows that it has become `request: 'C:Usersdl164OneDriveDocumentsdevempty-projectsrcApp.js.macro-fb0d4180853462d3bcb94480fb3b2b62c62519e0b35074e4a51d1c603706c075.css'` at which I got stuck trying to figure out where the dependencies were coming from. 

I verified that modifying the path [here ](https://github.com/devongovett/unplugin-parcel-macros/blob/d523d9f7c768050e83e5bedc503df75377a3d68a/unplugin-macros.js#L78)changed what the request field became, but attempting to use posix separators or further escaping all cause it to fail at an earlier step (not super sure about this but it no longer creates a [NormalModule](https://github.com/webpack/webpack/blob/46cee287199d7dce6879aa78046fae1df9a101e6/lib/NormalModule.js#L321) with a macro path that I was previously logging against):
```
ERROR in ./src/App.js 50:0-155
Module not found: Error: Can't resolve 'C:\Users\dl164\OneDrive\Documents\dev\empty-project\src\App.js.macro-fb0d4180853462d3bcb94480fb3b2b62c62519e0b35074e4a51d1c603706c075.css' in 'C:\Users\dl164\OneDrive\Documents\dev\empty-project\src'
```
```
Module not found: Error: Can't resolve 'C:/Users/dl164/OneDrive/Documents/dev/empty-project/src/App.js.macro-fb0d4180853462d3bcb94480fb3b2b62c62519e0b35074e4a51d1c603706c075.css' in 'C:\Users\dl164\OneDrive\Documents\dev\empty-project\src'
```

**Comment by alexander-akait** [comment]
You need to have `URL` in `import`, so if you want to use absolute path please use `file://` protocol

**Reply by LFDanLu** [reply]
Thanks for the suggestion! Just to make sure I understand, you are suggesting that the generated import statement should become
```
'import "file:\\\\C:\\Users\\dl164\\OneDrive\\Documents\\dev\\empty-project\\src\\App.js.macro-...
```
correct? When I tried this I got the following error instead:
```
WARNING in C:Usersdl164OneDriveDocumentsdevempty-projectsrcApp.js.macro-fb0d4180853462d3bcb94480fb3b2b62c62519e0b35074e4a51d1c603706c075.css 53:4-66:5
Module not found: Error: Can't resolve '!!../node_modules/css-loader/dist/cjs.js!C:Usersdl164OneDriveDocumentsdevempty-projectsrcApp.js.macro-fb0d4180853462d3bcb94480fb3b2b62c62519e0b35074e4a51d1c603706c075.css' in 'C:\Users\dl164\OneDrive\Documents\dev\empty-project'
```

Replacing the separators with the posix separators instead, aka a path that looks like 
``` 
'import "file://C:/Users/dl164/OneDrive/Documents/dev/empty-project/src/App.js.macro-fb0d4180853462d3bcb94480fb3b2b62c62519e0b35074e4a51d1c603706c075.css";',
```` 
yielded 
```
ERROR in ./src/App.js.macro-fb0d4180853462d3bcb94480fb3b2b62c62519e0b35074e4a51d1c603706c075.css 8:6-171
Module not found: Error: Can't resolve './App.js.macro-fb0d4180853462d3bcb94480fb3b2b62c62519e0b35074e4a51d1c603706c075.css' in 'C:\Users\dl164\OneDrive\Documents\dev\empty-project\src'
```

Would generating a relative import path be preferable?

**Reply by alexander-akait** [reply]
> Would generating a relative import path be preferable?

Yeah, always, anyway `file:` should work, can you show me a code/reproducible test repo when it broken?

**Reply by LFDanLu** [reply]
Sure, I've updated https://github.com/LFDanLu/s2-test-app so it has 
```
let id = `file:\\\\${filePath}.macro-${hash.digest('hex')}.${asset.type}`;
``` 
here: https://github.com/LFDanLu/s2-test-app/blob/main/patches/unplugin-parcel-macros%2B0.0.3.patch with some logging. Reverting to the [commit before](https://github.com/LFDanLu/s2-test-app/commit/3a1089933b0d7a99b398422de2e4ffe3ecd3a1a6) will generate the original error. The error reproduces via using Git Bash on Windows with node version 18.18.2

**Reply by alexander-akait** [reply]
Can you create reproducible test repo using github?

**Reply by LFDanLu** [reply]
Apologies if I'm misunderstanding the ask here, but https://github.com/LFDanLu/s2-test-app using the following steps should reproduce the issue:
1. Pull the project down and install.
2. Run using yarn start.
3. Note that it errors

**Reply by alexander-akait** [reply]
Bug in `unplugin-macros.js`, you can't use `file:\\\\${filePath}.macro-${hash.digest('hex')}.${asset.type}`, because it is wrong `file` URL, the right usage is:
```js
const { pathToFileURL } = require('node:url');
let id = pathToFileURL(`${filePath}.macro-${hash.digest('hex')}.${asset.type}`).toString();
```

anyway it will not work, because it is a virtual module, so you can't use `file` schema in such cases, there is not such file at all

**Reply by LFDanLu** [reply]
gotcha, thank you! I'll reach out to the maintainer of the `unplugin-parcel-macros` about this then.
